Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch

```diff
@@ -0,0 +1,147 @@
++                do {
++                    blockposition2.c(EnumDirection.DOWN);
++                } while (!iworldreader.getType(blockposition2).isAir());
++                do {
++                    blockposition2.c(EnumDirection.DOWN);
++                } while (iworldreader.getType(blockposition2).isAir() && blockposition2.getY() > 0);
++            }
+ 
+-            if (SpawnerCreature.a(EntityPositionTypes.Surface.ON_GROUND, iworldreader, blockposition2, EntityTypes.WANDERING_TRADER)) {
+-                blockposition1 = blockposition2;
++            if (SpawnerCreature.a(EntityPositionTypes.Surface.ON_GROUND, iworldreader, blockposition2.immutableCopy(), EntityTypes.WANDERING_TRADER)) {
++                blockposition1 = blockposition2.immutableCopy();
+                 break;
++                // Paper end
+             }
+         }
+ 
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 287db5203f3a4995c2b09bfd9fe17734538e7466..87ad87ef41064d7ec839aab850ac0dedb9eacf0c 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -123,7 +123,7 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
+ 
+     protected World(WorldDataMutable worlddatamutable, ResourceKey<World> resourcekey, final DimensionManager dimensionmanager, Supplier<GameProfilerFiller> supplier, boolean flag, boolean flag1, long i, org.bukkit.generator.ChunkGenerator gen, org.bukkit.World.Environment env, java.util.concurrent.Executor executor) { // Paper
+         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(((WorldDataServer) worlddatamutable).getName()); // Spigot
+-        this.paperConfig = new com.destroystokyo.paper.PaperWorldConfig((((WorldDataServer)worlddatamutable).getName()), this.spigotConfig); // Paper
++        this.paperConfig = new com.destroystokyo.paper.PaperWorldConfig((((WorldDataServer) worlddatamutable).getName()), this.spigotConfig, env); // Paper
+         this.chunkPacketBlockController = this.paperConfig.antiXray ? new ChunkPacketBlockControllerAntiXray(this, executor) : ChunkPacketBlockController.NO_OPERATION_INSTANCE; // Paper - Anti-Xray
+         this.generator = gen;
+         this.world = new CraftWorld((WorldServer) this, gen, env);
+diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
+index efe10b7f587810749038b67db14af0bf8b9b4374..d0b40afc9e503be352afc925a664b2207a480f7f 100644
+--- a/src/main/java/net/minecraft/server/WorldServer.java
++++ b/src/main/java/net/minecraft/server/WorldServer.java
+@@ -227,7 +227,24 @@ public class WorldServer extends World implements GeneratorAccessSeed {
+         this.L = new ObjectLinkedOpenHashSet();
+         this.Q = flag1;
+         this.server = minecraftserver;
+-        this.mobSpawners = list;
++        // Paper start - enable/disable MobSpawners per world
```

Comment:
This feels like plugin territory...

Reply:
Seems like it would need some hacky reflection to accomplish this with a plugin, given the list passed into the constructor is immutable

Reply:
Unless there's something I'm missing, this can be done with API only: https://papermc.io/javadocs/paper/1.16/com/destroystokyo/paper/event/entity/PreSpawnerSpawnEvent.html

Reply:
Ah, I believe that API is for AbstractMobSpawner / Mob Spawner Blocks, not the classes implementing MobSpawner, which are: MobSpawnerCat, MobSpawnerPatrol, MobSpawnerPhantom, MobSpawnerTrader, and MobSpawnerSiege

Reply:
Bless md's awful mappings then! Feel free to resolve this as I can't from the GH app I believe